### PR TITLE
[GH-1555] add null check to rename-gather-bulk

### PR DESCRIPTION
**Pull request by rfricke-asymmetrik**
### Purpose
During eMerge testing, an error occurred. This error was as a result of there being a null output in a fileref field. The null output was not specifically being handled when trying to create the request for the TDR sink and thus, an error was born.

https://broadinstitute.atlassian.net/browse/GH-1555

### Changes

- Added null check to `rename-gather-bulk`
- Added test case from null output

### Review Instructions

- Run unit test `test-rename-gather-bulk`

### CODE REVIEW

File: api/test/wfl/unit/sink_test.clj [hunk 1]
```diff
@@ -25,11 +25,13 @@
...
         result (sink/rename-gather-bulk (UUID/randomUUID) dataset "flowcell" inputs {:flowcell_tgz "flowcell_tgz"
-                                                                                     :flowcell_id "flowcell_id"})]
+                                                                                     :flowcell_id "flowcell_id"
+                                                                                     :sample_rename_map "sample_rename_map"})]
     (is (= (:flowcell_id inputs)
            (:flowcell_id result)))
     (is (= (:flowcell_tgz inputs)
-           (-> result :flowcell_tgz :sourcePath)))))
+           (-> result :flowcell_tgz :sourcePath)))
+    (is (= nil (:sample_rename_map result)))))
```
Comment (okotsopoulos) [inline api/test/wfl/unit/sink_test.clj:34]: Could instead write as:

```
(is (nil? (:sample_rename_map result)))
```

File: api/test/wfl/unit/sink_test.clj [hunk 2]
```diff
@@ -25,11 +25,13 @@
   (let [inputs (resources/read-resource "sarscov2_illumina_full/inputs.edn")
         dataset (resources/read-resource "sarscov2-illumina-full-inputs.json")
         result (sink/rename-gather-bulk (UUID/randomUUID) dataset "flowcell" inputs {:flowcell_tgz "flowcell_tgz"
-                                                                                     :flowcell_id "flowcell_id"})]
+                                                                                     :flowcell_id "flowcell_id"
+                                                                                     :sample_rename_map "sample_rename_map"})]
```
Comment (okotsopoulos) [inline api/test/wfl/unit/sink_test.clj:29]: Is this a fileref, where is this established if so?

**Review by okotsopoulos — APPROVED** [review]
Thanks for fixing!

Is this issue specific to the fileref type, or could null outputs of other types also pose issues?

If you also have available the specific conditions which lead to the error, I'd like to see them.

**Comment by rfricke-asymmetrik** [comment]
Thank you for the review. While the other fields should not have issues with nulls, I did add a check for `nil?` on the value in the rename-gather-bulk method as a nil value in any field other than fileref would have netted an "unknown operation" error.

The scenario that this was a fix for is like so:

```outputs: {:output_vcf_index "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-GtcToVcf/204126290052_R01C01.vcf.gz.tbi", 
   :arrays_subset_variant_calling_control_metrics_file nil,
   :genotype_concordance_summary_metrics_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-GenotypeConcordance/204126290052_R01C01.genotype_concordance_summary_metrics",
   :fingerprint_summary_metrics_file nil,
   :chip_well_barcode_params_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-CreateChipWellBarcodeParamsFile/cacheCopy/params.txt",
   :analysis_version_number_output 12894,
   :genotype_concordance_detail_metrics_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-GenotypeConcordance/204126290052_R01C01.genotype_concordance_detail_metrics",
   :gtc_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-AutoCall/204126290052_R01C01.gtc",
   :green_idat_md5_cloud_path "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-GreenIdatMd5Sum/v1_0110b839-6709-4c0f-80e2-8c6e526b22e3_7e089d9f-f9d6-4ff6-b7a7-2590ec120087.md5sum",
   :output_fingerprint_vcf_index "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-SelectFingerprintVariants/204126290052_R01C01.fingerprint.vcf.gz.tbi",
   :arrays_variant_calling_summary_metrics_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-CollectArraysVariantCallingMetrics/204126290052_R01C01.arrays_variant_calling_summary_metrics",
   :fingerprint_detail_metrics_file nil,
   :baf_regress_metrics_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-CreateBafRegressMetricsFile/204126290052_R01C01.bafregress_metrics",
   :output_fingerprint_vcf "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-SelectFingerprintVariants/204126290052_R01C01.fingerprint.vcf.gz", :reference_fingerprint_vcf_index nil,
   :arrays_subset_variant_calling_detail_metrics_file nil,
   :red_idat_md5_cloud_path "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-RedIdatMd5Sum/v1_0110b839-6709-4c0f-80e2-8c6e526b22e3_4c406c7e-b9b3-4a72-91e7-12dbab382971.md5sum",
   :chip_well_barcode_output "204126290052_R01C01",
   :output_vcf "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-GtcToVcf/204126290052_R01C01.vcf.gz",
   :reference_fingerprint_vcf nil,
   :arrays_variant_calling_control_metrics_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-CollectArraysVariantCallingMetrics/204126290052_R01C01.arrays_control_code_summary_metrics",
   :contamination_metrics_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-CreateVerifyIDIntensityContaminationMetricsFile/204126290052_R01C01.verifyidintensity_metrics",
   :genotype_concordance_contingency_metrics_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-GenotypeConcordance/204126290052_R01C01.genotype_concordance_contingency_metrics",
   :arrays_variant_calling_detail_metrics_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-CollectArraysVariantCallingMetrics/204126290052_R01C01.arrays_variant_calling_detail_metrics", :arrays_subset_variant_calling_summary_metrics_file nil,
   :output_fingerprint_json_file "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-VcfToMercuryFingerprintJson/204126290052_R01C01.fingerprint.json",
   :output_vcf_md5_cloud_path "gs://fc-e036248e-067b-4365-8c1e-5eb25103681f/1e6ef67d-7538-4831-aa96-e4667db550c8/Arrays/78e9af3e-9793-42de-aaf9-90611f34f054/call-IlluminaGenotypingArray/IlluminaGenotypingArray/5ca75b2d-4c56-4706-8c18-4a8627241e48/call-VcfMd5Sum/204126290052_R01C01.vcf.gz.md5sum"}

fromOutputs: {:output_vcf_index "output_vcf_index",
   :genotype_concordance_summary_metrics_file "genotype_concordance_summary_metrics_file",
   :fingerprint_summary_metrics_file "fingerprint_summary_metrics_file",
   :analysis_version_number_output "analysis_version_number_output",
   :genotype_concordance_detail_metrics_file "genotype_concordance_detail_metrics_file",
   :gtc_file "gtc_file",
   :arrays_variant_calling_summary_metrics_file "arrays_variant_calling_summary_metrics_file",
   :fingerprint_detail_metrics_file "fingerprint_detail_metrics_file",
   :baf_regress_metrics_file "baf_regress_metrics_file",
   :chip_well_barcode_output "chip_well_barcode_output",
   :output_vcf "output_vcf",
   :arrays_variant_calling_control_metrics_file "arrays_variant_calling_control_metrics_file",
   :last_modified_date "last_modified_date",
   :genotype_concordance_contingency_metrics_file "genotype_concordance_contingency_metrics_file",
   :arrays_variant_calling_detail_metrics_file "arrays_variant_calling_detail_metrics_file"}

table schema: {:name "ArraysOutputsTable"
                      :columns [{:name "chip_well_barcode_output"
                                 :datatype "text"
                                 :array_of false}
                                {:name "analysis_version_number_output"
                                 :datatype "integer"
                                 :array_of false}
                                {:name "gtc_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "arrays_variant_calling_control_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "arrays_variant_calling_detail_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "arrays_variant_calling_summary_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "baf_regress_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "fingerprint_detail_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "fingerprint_summary_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "genotype_concordance_contingency_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "genotype_concordance_detail_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "genotype_concordance_summary_metrics_file"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "last_modified_date"
                                 :datatype "datetime"
                                 :array_of false}
                                {:name "output_vcf"
                                 :datatype "fileref"
                                 :array_of false}
                                {:name "output_vcf_index"
                                 :datatype "fileref"
                                 :array_of false}]
                      :primaryKey ["chip_well_barcode_output" "analysis_version_number_output"]
                      :partitionMode "none"
                      :datePartitionOptions nil
                      :intPartitionOptions nil
                      :rowCount nil}```

File: api/test/wfl/unit/sink_test.clj (hunk 2 above)
Comment (rfricke-asymmetrik) [reply api/test/wfl/unit/sink_test.clj:29]: Sorry, back to this. Yes, it is a fileref. The dataset under `sarscov2-illumina-full-inputs.json` contains the table schema for `flowcell` that shows this field as a fileref field. The inputs did not make use of it, so I picked it to add a test for null handling on the fileref columns.
